proposed/resource-limit/resource-limit.md

```diff
@@ -0,0 +1,357 @@
+
+### Non-Goals
+
+TDB
+
+## Stakeholders and Reviewers
+
+- .NET team (System.Threading)
+- ASP.NET Core team
+
+## Design
+
+```c#
+    public abstract class ResourceLimiter
+    {
+        // an inaccurate view of resources
+        abstract long EstimatedCount { get; }
+
+        // Fast synchronous attempt to acquire resources
+        // Set requestedCount to 0 to get whether resource limit has been reached
+        abstract bool TryAcquire(long requestedCount, out Resource resource);
+
+        // Wait until the requested resources are available
+        // Set requestedCount to 0 to wait until resource is replenished
+        // If unsuccessful, throw
+        abstract ValueTask<Resource> AcquireAsync(long requestedCount, CancellationToken cancellationToken = default);
+    }
+
+    // Represent an aggregated resource (e.g. a resource limiter aggregated by IP), no identified use within dotnet/runtime yet
+    // Abstraction likely to be part of ASP.NET Core
+    public abstract class AggregatedResourceLimiter<TKey>
+    {
+        // an inaccurate view of resources
+        abstract long EstimatedCount(TKey resourceID);
+
+        // Fast synchronous attempt to acquire resources
+        // Set requestedCount to 0 to get whether resource limit has been reached
+        abstract bool TryAcquire(TKey resourceID, long requestedCount, out Resource resource);
+
+        // Wait until the requested resources are available
+        // Set requestedCount to 0 to wait until resource is replenished
+        // If unsuccessful, throw
+        abstract ValueTask<Resource> AcquireAsync(TKey resourceID, long requestedCount, CancellationToken cancellationToken = default);
+    }
+
+    public struct Resource : IDisposable
+    {
+        public object? State { get; init; }
+
+        private Action<Resource>? _onDispose;
+
+        public Resource(object? state, Action<Resource>? onDispose)
+        {
+            State = state;
+            _onDispose = onDispose;
+        }
+
+        public void Dispose()
+        {
+            _onDispose?.Invoke(this);
+        }
+
+        public static Resource NoopResource = new Resource(null, null);
+    }
+```
+
+The `Resource` struct is designed to handle the releasing of the resource. This way, the user won't be able to release more resources than was obtained, unlike Semaphores. We've also identified scenarios where additional information need to be returned by the limiter, such as reason phrases, response codes, percentage of rate saturation, retry after etc. For these cases, the object `Resource.State` can be used to store the additional metadata.
```

Comment:
> We've also identified scenarios where additional information need to be returned by the limiter, such as reason phrases, response codes, percentage of rate saturation, retry after etc. For these cases, the object `Resource.State` can be used to store the additional metadata.

This scenario doesn't fit with the proposed `bool TryAcquire(long requestedCount, out Resource resource);` API. If a `bool TryXXX` method returns `false`, it isn't a canonical pattern that the caller can use the `out` parameter. You tried to get it, but it failed. In this case, how can information be passed back to the caller in order to tell them the reason or "retry after"?

Reply:
Ah good point. I was thinking that even when `TryAcquire` returns false, it can choose to return a `Resource`. It's just that it's guaranteed to not be null when `TryAcquire` returns true.